torch/distributed/distributed_c10d.py

```diff
@@ -1408,47 +1468,63 @@ def recv(tensor: torch.Tensor, src: Optional[int] = None, group: Optional[Proces
         return src
 
 
-class P2POp:
+@contextlib.contextmanager
+def _coalescing_manager(group: Optional[ProcessGroup] = None, device: Optional[torch.device] = None, reqs: Optional[List] = None):
     """
-    A class to build point-to-point operations for ``batch_isend_irecv``.
-
-    This class builds the type of P2P operation, communication buffer, peer rank,
-    Process Group group, and tag. Instances of this class will be passed to
-    ``batch_isend_irecv`` for point-to-point communications.
+    A context manager used to coalesce collectives or P2P operations when possible.
 
     Args:
-        op (Callable): A function to send data to or receive data from a peer process.
-            The type of ``op`` is either ``torch.distributed.isend`` or
-            ``torch.distributed.irecv``.
-        tensor (Tensor): Tensor to send or receive.
-        peer (int): Destination or source rank.
-        group (ProcessGroup, optional): The process group to work on. If None,
+        group (`ProcessGroup`, optional): The process group to work on. If None,
             the default process group will be used.
-        tag (int, optional): Tag to match send with recv.
-    """
+        device (`torch.device`, optional): Default is None, set to a device if there isn't a `**_coalesced` implementation
+            by the backend.
+        reqs (`List`, optional): if provided, put created `Work` items in the `reqs` list. If `async_op` is False, no
+            item will be put in the list. The number of created `Work` items may vary by backend.
 
-    def __init__(self, op, tensor, peer, group=None, tag=0):
-        self.op = op
-        self.tensor = tensor
-        self.peer = peer
-        self.group = group
-        self.tag = tag
-
-    def __new__(cls, op, tensor, peer, group=None, tag=0):
-        _check_op(op)
-        _check_single_tensor(tensor, "tensor")
-        return object.__new__(cls)
-
-
-@contextlib.contextmanager
-def _coalescing_manager(group, device, reqs):
+    Examples:
+        >>> # xdoctest: +SKIP("no rank")
+        >>> reqs = []
+        >>> with _coalescing_manager(reqs=reqs):
+        >>>     for i in range(num_coll):
+        >>>         dist.all_reduce(tensors[i], async_op=True)
+        >>> for req in reqs:
+        >>>     req.wait()
+    """
     if group is None:
         group = _get_default_group()
-    group._start_coalescing(device)
+    op_list = _world.pg_coalesce_state.setdefault(group, [])
```

Comment:
do we plan to make this work for multi-thread PG? If so, even if there is GIL, I assume we still need a lock here?

Reply:
I guess I have no such plan for now for real use case.
But if the lack of support causes our CI to fail, I can look into that.